### PR TITLE
#3668 - SIMS Beta User Validation

### PR DESCRIPTION
**Pull request by andrepestana-aot**
image

- Created a new env var ALLOW_BETA_USERS_ONLY to enable/disable the period of beta users;
- Added table `sims.beta_users_authorizations`;
- Added error message to student's login page;
- Added conditions to allow only BCSC student users to login if they are registered in `sims.beta_users_authorizations`;
- Added tests to check if the system is enabling beta users only and if the user is registered in `sims.beta_users_authorizations`; 

Migration revert:
image

### CODE REVIEW

File: .github/workflows/release-deploy-all.yml [hunk 1]
```diff
@@ -97,6 +97,7 @@ env:
   TLS_CERTIFICATE: ${{ secrets.TLS_CERTIFICATE }}
   TLS_KEY: ${{ secrets.TLS_KEY }}
   TLS_CA_CERTIFICATE: ${{ secrets.TLS_CA_CERTIFICATE }}
+  ALLOW_BETA_USERS_ONLY: ${{ secrets.ALLOW_BETA_USERS_ONLY }}
```
Comment (andrewsignori-aot) [inline .github/workflows/release-deploy-all.yml:100]: Please use `vars` instead for non-sensitive data.

File: sources/packages/backend/apps/api/src/services/user/user.model.ts [hunk 2]
```diff
@@ -5,4 +5,6 @@ export interface UserLoginInfo {
   isActive: boolean;
   studentId?: number;
   identityProviderType?: IdentityProviders;
+  firstName?: string;
+  lastName: string;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/user/user.model.ts:9]: Where are these properties consumed?

File: sources/packages/backend/apps/api/src/services/user/user.service.ts [hunk 3]
```diff
@@ -140,4 +157,30 @@ export class UserService extends DataModelService<User> {
...
+      }),
+      lastName: Raw((alias) => `LOWER(${alias}) = LOWER(:lastName)`, {
+        lastName,
+      }),
+      enabledFrom: LessThanOrEqual(now),
+    };
+    if (!givenNames) {
+      where.givenNames = IsNull();
+    }
+    return await this.betaUsersAuthorizationsRepo.exists({ where });
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/user/user.service.ts:184]: No need for `await`.

File: sources/packages/backend/apps/db-migrations/src/migrations/1726775504621-CreateBetaUsersAuthorizationsTable.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        "Create-beta-users-authorizations-table.sql",
+        "BetaUsersAuthorizations",
+      ),
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      getSQLFileData(
+        "Drop-beta-users-authorizations-table.sql",
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/migrations/1726775504621-CreateBetaUsersAuthorizationsTable.ts:19]: Please rename it to `Rollback-create-beta-users-authorizations-table.sql`

File: sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE IF NOT EXISTS beta_users_authorizations(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql:1]: Please remove the `IF NOT EXISTS`.

Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql:1]: Please add the `sims` schema.

File: sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  id SERIAL PRIMARY KEY,
+  given_names VARCHAR(100) NULL,
+  last_name VARCHAR(100) NOT NULL,
+  enabled_from TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+  CONSTRAINT beta_users_authorizations_name_unique UNIQUE (given_names, last_name)
+);
+
+-- ## Comments
+COMMENT ON TABLE beta_users_authorizations IS 'Beta users authorizations table. Only users with a BCSC credential that matches given name and last name in the period starting from `enabled_from` column will be allowed during the beta user period.';
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql:10]: The back ticks will not affect the format. Please change it to single quotes or removed it (`'enabled_from'`).

Comment (andrewsignori-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/BetaUsersAuthorizations/Create-beta-users-authorizations-table.sql:10]: Please add the `sims` schema in all the comments.

File: sources/packages/web/src/services/AuthService.ts [hunk 7]
```diff
@@ -258,6 +266,13 @@ export class AuthService {
         await this.executeSiteminderLogoff(redirectUri);
         break;
       }
+      case ClientIdType.Student: {
+        if (options?.invalidBetaUser) {
+          redirectUri += "/login/invalid-beta-user";
+        }
+        await this.executeSiteminderLogoff(redirectUri);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/AuthService.ts:273]: `executeSiteminderLogoff` is only applicable for IDIR and BCeID. It should not be invoked for BCSC.
We need to check if the student is authenticated using BCeID before calling `executeSiteminderLogoff`.

File: sources/packages/web/src/views/student/Login.vue [hunk 8]
```diff
@@ -20,6 +20,16 @@
                   Whether you are a new or returning user—log in or register
                   using the BC Services Card account.
                 </p>
+                <v-row>
+                  <v-col>
+                    <banner
+                      v-if="errorMessage"
+                      class="my-2"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/Login.vue:27]: The banner has too much space on the top and it is different from the bottom.
Can we reduce the top spacing to be more like the bottom?
image

File: sources/packages/web/src/views/student/Login.vue [hunk 9]
```diff
@@ -20,6 +20,16 @@
                   Whether you are a new or returning user—log in or register
                   using the BC Services Card account.
                 </p>
+                <v-row>
+                  <v-col>
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/Login.vue:24]: Is there a need for a `v-row` and a `v-col` here?

File: sources/packages/backend/apps/api/src/services/user/user.model.ts (hunk 2 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/services/user/user.model.ts:9]: Yes, they aren't consumed anywhere. I had added they won't be used. Thanks!

File: sources/packages/backend/apps/api/test/auth.e2e-spec.ts
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/test/auth.e2e-spec.ts:1]: This file is pretty old and his place right makes the imports come from `..src/`.
Can you please move the file inside the folder "src/auth/_tests_/e2e"?

File: sources/packages/backend/apps/api/test/auth.e2e-spec.ts [hunk 10]
```diff
@@ -69,6 +99,100 @@ describe("Authentication (e2e)", () => {
...
+      "user is registered in beta users authorizations table.",
+    async () => {
+      // Arrange
+      Object.defineProperty(configService, "allowBetaUsersOnly", {
+        value: true,
+        writable: true,
+      });
+
+      const jwtService = new JwtService();
+      const decodedToken = jwtService.decode(studentAccessToken);
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/test/auth.e2e-spec.ts:145]: `givenNames` and `lastName` can be decoded at `beforeAll`.

File: sources/packages/backend/apps/api/src/services/user/user.service.ts [hunk 11]
```diff
@@ -1,16 +1,29 @@
...
+  UpdateResult,
+} from "typeorm";
 import {
   DataModelService,
   IdentityProviders,
   Student,
   User,
 } from "@sims/sims-db";
 import { UserLoginInfo } from "./user.model";
+import { BetaUsersAuthorizations } from "@sims/sims-db/entities/beta-users-authorizations.model";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/user/user.service.ts:17]: It should be consumed from an index.

**Review by andrewsignori-aot — COMMENTED** [review]
Great work. Please take a look at the comments.

File: devops/Makefile [hunk 12]
```diff
@@ -413,6 +413,7 @@ deploy-api:
 		-p TLS_CERTIFICATE=$(TLS_CERTIFICATE) \
 		-p TLS_KEY=$(TLS_KEY) \
 		-p TLS_CA_CERTIFICATE=$(TLS_CA_CERTIFICATE) \
+    -p ALLOW_BETA_USERS_ONLY=$(ALLOW_BETA_USERS_ONLY) \
```
Comment (guru-aot) [inline devops/Makefile:416]: 👍

File: sources/packages/backend/apps/api/test/auth.e2e-spec.ts
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/test/auth.e2e-spec.ts:1]: Just a small correction, the tests folder has an underscore before and after. It was converted to italic in the comment.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrepestana-aot

**Pull request merged**